### PR TITLE
cache: Add `.Set` and `.Add` methods to cache clients

### PR DESCRIPTION
**Pull request by 56quarters**
**What this PR does**:

This change adds a synchronous version of `.Set` to Memcached and Redis clients as well as the various `Cache` wrapper implementations. This allows callers to set a key and be sure it exists in the cache. This change also adds an `.Add` method which conditionally adds an item to the cache only if it does not already exist.

**Which issue(s) this PR fixes**:

This change is a prerequisite for grafana/mimir#9386

**Checklist**
- [x] Tests updated
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: cache/client.go [hunk 1]
```diff
@@ -17,6 +17,7 @@ import (
 // Common functionality shared between the Memcached and Redis Cache implementations
 
 const (
+	opAdd            = "add"
```
Comment (dimitarvdimitrov) [inline cache/client.go:20]: this isn't used to initialize `cm.operations`

File: cache/lru.go [hunk 2]
```diff
@@ -103,6 +103,41 @@ func (l *LRUCache) SetMultiAsync(data map[string][]byte, ttl time.Duration) {
...
+	return err
+}
+
+func (l *LRUCache) Add(ctx context.Context, key string, value []byte, ttl time.Duration) error {
+	err := l.c.Add(ctx, key, value, ttl)
+
+	// When a caller uses the Add method, the presence of absence of an entry in the cache
+	// has significance. In order to maintain the semantics of that, we only add an entry to
+	// the LRU when it was able to be successfully added to the shared cache.
+	if err == nil {
```
Comment (dimitarvdimitrov) [inline cache/lru.go:127]: i didn't think of this. Can you add a tiny test for it?

File: cache/memcached_client.go [hunk 3]
```diff
@@ -28,6 +28,7 @@ import (
 
 const (
 	dnsProviderUpdateInterval = 30 * time.Second
+	maxTTL                    = 30 * 24 * time.Hour
```
Comment (dimitarvdimitrov) [inline cache/memcached_client.go:31]: where does this come from? number of seconds in an in32? hmm, no, that's way more.

File: cache/memcached_client.go [hunk 4]
```diff
@@ -339,6 +374,20 @@ func (c *MemcachedClient) setSingleItem(key string, value []byte, ttl time.Durat
 	})
 }
 
+// TODO: Docs
```
Comment (dimitarvdimitrov) [inline cache/memcached_client.go:377]: 👀

**Comment by 56quarters** [comment]
> nice one, I'm not sure about the 30d TTL, otherwise LGTM

When you give Memcached a TTL longer than 30d, it treats the number of seconds as a Unix timestamp. E.g. if you give it a 31d TTL your item will expire on February 1st 1970 -- immediately.

File: cache/memcached_client.go (hunk 4 above)
Comment (56quarters) [reply cache/memcached_client.go:377]: :facepalm: I'll fix

File: cache/client.go (hunk 1 above)
Comment (56quarters) [reply cache/client.go:20]: Fixed

File: cache/lru.go (hunk 2 above)
Comment (56quarters) [reply cache/lru.go:127]: Good idea, will do :+1:

**Pull request merged**